rust/src/two_stacks_lite/mod.rs

```diff
@@ -0,0 +1,82 @@
+use std::collections::VecDeque;
+
+use alga::general::AbstractMagma;
+use alga::general::Identity;
+
+use crate::FifoWindow;
+use crate::ops::AggregateOperator;
+use crate::ops::AggregateMonoid;
+
+#[derive(Clone)]
+pub struct TwoStacksLite<BinOp>
+where
+    BinOp: AggregateMonoid<BinOp> + AggregateOperator + Clone
+{
+    queue: VecDeque<BinOp::Partial>,
+    agg_back: BinOp::Partial,
+    front_len: usize,
+}
+
+impl<BinOp> FifoWindow<BinOp> for TwoStacksLite<BinOp>
+where
+    BinOp: AggregateMonoid<BinOp> + AggregateOperator + Clone
+{
+    fn new() -> Self {
+        Self {
+            queue: VecDeque::new(),
+            agg_back: BinOp::Partial::identity(),
+            front_len: 0,
+        }
+    }
+    fn name() -> &'static str {
+        "two_stacks_lite"
+    }
+    fn push(&mut self, v: BinOp::In) {
+        let lifted = BinOp::lift(v);
+        self.queue.push_back(lifted.clone());
+        self.agg_back = self.agg_back.operate(&lifted);
+    }
+    fn pop(&mut self) -> Option<BinOp::Out> {
+        if self.queue.is_empty() {
+            None
+        } else {
+            let end = self.queue.len() - 1;
+            if self.front_len == 0 {
+                let mut accum = BinOp::Partial::identity();
+                for partial in self.queue.iter_mut().rev() {
+                    accum = accum.operate(&partial);
```

Comment:
For non-commutative aggregation operators, this needs to be the other way around:
```rust
accum = partial.operate(accum);
```

Reply:
Good catch, fixed.

Reply:
This looks good now.